### ISSUE TITLE
"Find in Files" is initially unresponsive when searching a large folder structure

### ISSUE THREAD

**Issue by UweKeim** [notepad-plus-plus/notepad-plus-plus#11774]
### Description of the Issue

When using the “Find in Files” dialog to search through a large folder structure, the dialog becomes unresponsive right after starting the searching by clicking the “Find All” button.

<img width="432" alt="image" src="https://user-images.githubusercontent.com/415439/172814057-4fec49dc-be1e-4c6c-9024-2752ec654f18.png">

In my opinion this seems to be the case because when starting the searching first a directory tree seems to be read fully into memory before actually starting to search at all.

For me this is a bug.

I would love to have a “Find in Files” dialog that is responsive in every stage of the search. I.e. even when initially reading in the the folder structure and not only when later searching inside the actual files.

### Steps to Reproduce the Issue

1. Open the "Find in Files" dialog
2. Set the "Directory" field to a folder with lots of files and/or sub-folders
3. Click the "Find all" button

### Expected Behavior

The "Find in Files" dialog should start searching/scanning the folders and files and stay responsive (i.e. redraw itself, be able to handle user inputs like canceling the find).

### Actual Behavior

The "Find in Files" dialog remains unresponsive for several seconds up to a lot of seconds or even minutes, depending on the machine power and the size of the folder structure to search in.

### Debug Information

Notepad++ v8.4.2   (64-bit)
Build time : May 29 2022 - 16:47:30
Path : C:\Program Files\Notepad++\notepad++.exe
Command Line : $COMMAND_LINE_PLACEHOLDER$
Admin mode : OFF
Local Conf mode : OFF
Cloud Config : OFF
OS Name : Windows 11 (64-bit) 
OS Version : 21H2
OS Build : 22000.708
Current ANSI codepage : 1252
Plugins : 
    mimeTools (2.8)
    NppConverter (4.4)
    NppExport (0.4)

**Comment by alankilborn** [comment]
Related Community link: https://community.notepad-plus-plus.org/topic/23055

**Comment by donho** [comment]
@UweKeim 
Find in files not responsive problem is rather on the 2nd (and after) launch, but not the very first launch.
Could you confirm me the description above?

**Comment by donho** [comment]
@UweKeim 
When I said "the 1st time", I meant with Notepad++ fresh launch, the 1st time you launch Find in files.
Could you close your Notepad++, launch it again then run the operation Find in Files?

**Comment by UweKeim** [comment]
Yes, I did understand. As I said, it is just the opposite as what you described. To me it is:

- When first fresh start of NP++ and then Find in Files, the UI becomes unresponsive for quite a while, depending on the folder size.
- When then entering the Find in Files for the very same folder for a second time, the unresponsiveness is shorter. My assumption is that it uses some caching to access the data, probably some kind of OS cache, not NP++ cache.

Steps I performed:

1. Start NP++.
2. Open Find in Files.
3. Enter parameters, pointing the "Directory" field to a larger sub folder tree.
4. Click the "Find all" button.

It then starts stalling and after some longer time, the progress dialog starts showing up.

Then I continue:

1. Wait until the find results are shown in the panel in the main window and the Find in Files dialog is closed.
2. Open Find in Files again.
3. Leave the "Directory" field as to what it was set before..
4. Click the "Find all" button.

Now the stalling is much shorter, before the progress dialog appears.

------

To me that totally behaves like a two step process (I haven't looked at the NP++ sources though):

1. NP++ builds a directory tree in memory to know where to look for files.
2. NP++ starts processing the built in-memory directory tree.

Maybe I am completely wrong, but if not, it strongly looks to me that step 1 is UI-blocking whereas step 2 has a dedicated progress dialog.

If I am correct, I do think that the code that builds the directory tree should not build it completely in memory before starting to process the files, but rather start processing at the start folder entered by the user, and then iterate it "on the fly", processing both the folders _and_ the files it finds.

This would eliminate the long startup process since there is no need to wait until a full directory tree is built in-memory beforehand.

**Comment by Yaron10** [comment]
@donho,

After https://github.com/notepad-plus-plus/notepad-plus-plus/issues/11814, I've tested the case here.

I've searched for `color` in my NPP installation folder (portable, two sub folders containing x32 and x64), and set `*.xml` as a filter.

After ~3 seconds, I got 
![תמונה](https://user-images.githubusercontent.com/10229320/174450345-dd5015ca-7bee-4afd-811c-63396d610019.png)

No difference between first/next Find in files launches.
The issue exists in v8.2.0 as well.

**Comment by UweKeim** [comment]
Try searching a folder structure with hundreds and thousands of (sometimes deeply nested) sub folders.

In my experience: The more folders, the longer the stalling until the progress dialog starts.

**Comment by Yaron10** [comment]
@UweKeim,

Thank you for the detailed report.
